sdk/storage/Azure.Storage.Common/tests/Shared/ClientBuilder.cs

```diff
@@ -0,0 +1,178 @@
+        /// <param name="getServiceClientTokenCredential">Constructs a service client from the given arguments.</param>
+        /// <param name="getServiceClientAzureSasCredential">Constructs a service client from the given arguments.</param>
+        /// <param name="instrumentClient">Delegate to the test environment's InstrumentClient functionality.</param>
+        public ClientBuilder(
+            TenantConfigurationBuilder tenantConfigurationBuilder,
+            GetServiceClient getServiceClient,
+            GetServiceClientStorageSharedKeyCredential getServiceClientStorageSharedKeyCredential,
+            GetServiceClientTokenCredential getServiceClientTokenCredential,
+            GetServiceClientAzureSasCredential getServiceClientAzureSasCredential,
+            GetServiceClientOptions getServiceClientOptions)
+        {
+            Tenants = tenantConfigurationBuilder;
+            _getServiceClient = getServiceClient;
+            _getServiceClientStorageSharedKeyCredential = getServiceClientStorageSharedKeyCredential;
+            _getServiceClientTokenCredential = getServiceClientTokenCredential;
+            _getServiceClientAzureSasCredential = getServiceClientAzureSasCredential;
+            _getServiceClientOptions = getServiceClientOptions;
+        }
+
+        public TServiceClient GetServiceClient_SharedKey(TServiceClientOptions options = default)
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigDefault, options);
+
+        public TServiceClient GetServiceClient_SecondaryAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigSecondary);
+
+        public TServiceClient GetServiceClient_PreviewAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPreviewBlob);
+
+        public TServiceClient GetServiceClient_PremiumBlobAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumBlob);
+        public TServiceClient GetServiceClient_OAuth() =>
+            GetServiceClientFromOauthConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_OAuthAccount_SharedKey() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_ManagedDisk() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigManagedDisk);
+
+        public TServiceClient GetServiceClient_PremiumBlob() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumBlob);
+
+        public TServiceClient GetServiceClient_PremiumFile() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumFile);
+
+        public TServiceClient GetServiceClient_Hns() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigHierarchicalNamespace);
+
+        public TServiceClient GetServiceClient_SoftDelete() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigSoftDelete);
```

Comment:
I think these should live in service specific packages. Why Queues would be coupled to HNS account for example?
Btw. instead of embedding all the tenants inside this class we could make `ClientBuilder(Tenant)` and then for example in the `BlobTestBase` have `ClientBuilder(Tenants.TestConfigDefault)`  ,  `ClientBuilder(Tenants.TestConfigSecondary)` and so on, multiple instances and each working with some tenant.

Reply:
I've split these out into service-specific extensions.

I'm trying really hard not to scope creep this PR and so I'd rather not do the separate `ClientBuilder` per tenant just yet. I do think what we have now is still a noticeable improvement to our current infra.

Reply:
can you log work item for remaining part?

Reply:
#24457